File: spacetraders/src/responses.rs
use crate::shared;
use serde::Deserialize;
use chrono::{DateTime, Utc};

#[derive(Deserialize, Debug, Clone)]
pub struct GameStatus {
    pub status: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct UserInfoData {
    pub username: String,
    pub credits: i32,
    pub ships: Vec<shared::Ship>,
    pub loans: Vec<shared::Loan>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct UserInfo {
    pub user: UserInfoData,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AvailableLoan {
    #[serde(rename = "type")]
    pub loan_type: shared::LoanType,
    pub amount: i32,
    pub rate: f64,
    #[serde(rename = "termInDays")]
    pub term_in_days: i32,
    #[serde(rename = "collateralRequired")]
    pub collateral_required: bool,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AvailableLoans {
    pub loans: Vec<AvailableLoan>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ShipsForSale {
    pub ships: Vec<shared::ShipForSale>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct PurchaseOrder {
    pub credits: i32,
    pub order: Vec<shared::Order>,
    pub ship: shared::Ship,
}

#[derive(Deserialize, Debug, Clone)]
pub struct AvailableLocations {
    pub locations: Vec<shared::Location>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct FlightPlan {
    #[serde(rename = "flightPlan")]
    pub flight_plan: shared::FlightPlanData,
}

#[derive(Deserialize, Debug, Clone)]
pub struct SystemsInfo {
    pub systems: Vec<shared::SystemsInfoData>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct User {
    pub id: String,
    pub username: String,
    pub picture: Option<String>,
    pub email: Option<String>,
    pub credits: i32,
    #[serde(rename = "createdAt")]
    pub created_at: DateTime<Utc>,
    #[serde(rename = "updatedAt")]
    pub updated_at: DateTime<Utc>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct ClaimUsernameResponse {
    pub token: String,
    pub user: User,
}

#[derive(Deserialize, Debug, Clone)]
pub struct YourShips {
    pub ships: Vec<shared::Ship>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct LocationInfo {
    pub planet: shared::SystemsInfoLocation,
}

#[derive(Deserialize, Debug, Clone)]
pub struct LoanInfo {
    pub loans: Vec<shared::Loan>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct PlanetMarketplace {
    pub planet: shared::PlanetMarketplaceData,
}

File: spacetraders/responses.go
package spacetraders

import "time"

type GetMyIpAddressResponse struct {
	Ip string `json:"ip"`
}

type GameStatusResponse struct {
	Status string `json:"status"`
}

type ClaimUsernameResponseUser struct {
	Username string `json:"username"`
	Credits  int    `json:"credits"`
	Ships    []Ship `json:"ships"`
	Loans    []Loan `json:"loans"`
}

type ClaimUsernameResponse struct {
	Token string                    `json:"token"`
	User  ClaimUsernameResponseUser `json:"user"`
}

type UserInfoUser struct {
	Username       string    `json:"username"`
	Credits        int       `json:"credits"`
	ShipCount      int       `json:"shipCount"`
	StructureCount int       `json:"structureCount"`
	JoinedAt       time.Time `json:"joinedAt"`
}

type GetMyInfoResponse struct {
	User UserInfoUser `json:"user"`
}

type GetFlightPlanResponse struct {
	FlightPlan FlightPlan `json:"flightPlan"`
}

type CreateFlightPlanResponse struct {
	FlightPlan FlightPlan `json:"flightPlan"`
}

type GetMyLoansResponse struct {
	Loans []Loan `json:"loans"`
}

type PayOffLoanResponse struct {
	Credits int    `json:"credits"`
	Loans   []Loan `json:"loans"`
}

type CreateLoanResponse struct {
	Credits int  `json:"credits"`
	Loan    Loan `json:"loan"`
}

type GetLocationResponse struct {
	Location    LocationDetails `json:"location"`
	DockedShips int             `json:"dockedShips"`
}

type GetLocationMarketplaceResponse struct {
	Marketplace []MarketplaceData
}

type CreatePurchaseOrderResponse struct {
	Credits int       `json:"credits"`
	Order   OrderData `json:"order"`
	Ship    Ship      `json:"ship"`
}

type CreateSellOrderResponse struct {
	Credits int       `json:"credits"`
	Order   OrderData `json:"order"`
	Ship    Ship      `json:"ship"`
}

type PurchaseShipResponse struct {
	Credits int  `json:"credits"`
	Ship    Ship `json:"ship"`
}

type GetMyShipRequest struct {
	Ship Ship `json:"ship"`
}

type GetMyShipsResponse struct {
	Ships []Ship `json:"ships"`
}

type JettisonCargoResponse struct {
	ShipId            string `json:"shipId"`
	Good              string `json:"good"`
	QuantityRemaining int    `json:"quantityRemaining"`
}

type PurchaseLocation struct {
	System   string `json:"system"`
	Location string `json:"location"`
	Price    int    `json:"price"`
}

type ShipForSale struct {
	ShipType          string             `json:"type"`
	Class             string             `json:"class"`
	MaxCargo          int                `json:"maxCargo"`
	LoadingSpeed      int                `json:"loadingSpeed"`
	Speed             int                `json:"speed"`
	Manufacturer      string             `json:"manufacturer"`
	Plating           int                `json:"plating"`
	Weapons           int                `json:"weapons"`
	PurchaseLocations []PurchaseLocation `json:"purchaseLocations"`
	RestrictedGoods   []string           `json:"restrictedGoods"`
}

type GetShipsForSaleResponse struct {
	ShipsForSale []ShipForSale `json:"ships"`
}

type SystemLocation struct {
	Symbol             string   `json:"symbol"`
	SystemLocationType string   `json:"type"`
	Name               string   `json:"name"`
	X                  int      `json:"x"`
	Y                  int      `json:"y"`
	AllowsConstruction bool     `json:"allowsConstruction"`
	Traits             []string `json:"traits"`
	Messages           []string `json:"messages"`
}

type GetSystemLocationsResponse struct {
	Locations []SystemLocation `json:"locations"`
}

type AvailableLoan struct {
	LoanType           string  `json:"type"`
	Amount             int     `json:"amount"`
	Rate               float64 `json:"rate"`
	TermInDays         int     `json:"termInDays"`
	CollateralRequired bool    `json:"collateralRequired"`
}

type GetAvailableLoansResponse struct {
	Loans []AvailableLoan `json:"loans"`
}

type WarpJumpResponse struct {
	FlightPlan FlightPlan `json:"flightPlan"`
}
